File: python/networking/p2p_file_sharing.py
"""P2P File Sharing App — decentralized file exchange between peers."""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional
import hashlib
import math


@dataclass
class FileChunk:
    """A chunk of a file, identified by index and hash."""
    index: int
    data: bytes
    sha256: str

    @staticmethod
    def from_data(index: int, data: bytes) -> "FileChunk":
        return FileChunk(
            index=index,
            data=data,
            sha256=hashlib.sha256(data).hexdigest(),
        )

    def verify(self) -> bool:
        return hashlib.sha256(self.data).hexdigest() == self.sha256


@dataclass
class SharedFile:
    """Metadata for a file shared by a peer."""
    name: str
    size: int
    sha256: str
    chunk_size: int = 4096
    chunk_count: int = 0

    def __post_init__(self) -> None:
        if self.chunk_count == 0:
            self.chunk_count = math.ceil(self.size / self.chunk_size) if self.size > 0 else 0


@dataclass
class PeerInfo:
    """Information about a known peer."""
    peer_id: str
    address: str
    port: int
    last_seen: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    shared_files: list[SharedFile] = field(default_factory=list)

    def is_stale(self, timeout_seconds: int = 300) -> bool:
        age = (datetime.now(timezone.utc) - self.last_seen).total_seconds()
        return age > timeout_seconds


@dataclass
class TransferRequest:
    """A request to transfer a file between peers."""
    request_id: str
    filename: str
    requester_id: str
    provider_id: str
    status: str = "pending"  # pending, active, completed, failed
    chunks_received: int = 0
    total_chunks: int = 0
    started: Optional[datetime] = None
    completed: Optional[datetime] = None

    @property
    def progress(self) -> float:
        if self.total_chunks == 0:
            return 0.0
        return self.chunks_received / self.total_chunks

    @property
    def progress_pct(self) -> str:
        return f"{self.progress * 100:.1f}%"


class PeerNode:
    """A peer in the P2P network. Manages local files, known peers,
    discovery, catalog exchange, and chunked file transfers."""

    def __init__(self, peer_id: str, address: str = "127.0.0.1", port: int = 0) -> None:
        self.peer_id = peer_id
        self.address = address
        self.port = port
        self.peers: dict[str, PeerInfo] = {}
        self.local_files: dict[str, bytes] = {}
        self.shared_catalog: dict[str, SharedFile] = {}
        self.transfers: dict[str, TransferRequest] = {}
        self._chunks_store: dict[str, list[Optional[FileChunk]]] = {}
        self._next_request_id = 0

    # --- Local file management ---

    def add_file(self, name: str, data: bytes, chunk_size: int = 4096) -> SharedFile:
        """Add a file to the local store and share it on the network."""
        self.local_files[name] = data
        sf = SharedFile(
            name=name,
            size=len(data),
            sha256=hashlib.sha256(data).hexdigest(),
            chunk_size=chunk_size,
        )
        self.shared_catalog[name] = sf
        return sf

    def remove_file(self, name: str) -> None:
        """Stop sharing a file."""
        self.local_files.pop(name, None)
        self.shared_catalog.pop(name, None)

    def get_chunk(self, filename: str, chunk_index: int) -> Optional[FileChunk]:
        """Retrieve a specific chunk of a local file (called by requesting peers)."""
        data = self.local_files.get(filename)
        if data is None:
            return None
        sf = self.shared_catalog.get(filename)
        if sf is None:
            return None
        start = chunk_index * sf.chunk_size
        end = min(start + sf.chunk_size, len(data))
        if start >= len(data):
            return None
        return FileChunk.from_data(chunk_index, data[start:end])

    # --- Peer discovery ---

    def discover_peer(self, peer_id: str, address: str, port: int) -> PeerInfo:
        """Register a known peer (from bootstrap, broadcast, or introduction)."""
        info = PeerInfo(peer_id=peer_id, address=address, port=port)
        self.peers[peer_id] = info
        return info

    def remove_peer(self, peer_id: str) -> None:
        """Remove a peer from the known list."""
        self.peers.pop(peer_id, None)

    def prune_stale_peers(self, timeout_seconds: int = 300) -> list[str]:
        """Remove peers not seen within the timeout. Returns removed IDs."""
        stale = [pid for pid, p in self.peers.items() if p.is_stale(timeout_seconds)]
        for pid in stale:
            del self.peers[pid]
        return stale

    def heartbeat(self, peer_id: str) -> None:
        """Update last-seen timestamp for a peer."""
        if peer_id in self.peers:
            self.peers[peer_id].last_seen = datetime.now(timezone.utc)

    # --- Catalog exchange ---

    def get_catalog(self) -> list[SharedFile]:
        """Return this peer's shared file catalog (sent to other peers)."""
        return list(self.shared_catalog.values())

    def receive_catalog(self, peer_id: str, catalog: list[SharedFile]) -> None:
        """Receive and store another peer's catalog."""
        if peer_id in self.peers:
            self.peers[peer_id].shared_files = catalog
            self.peers[peer_id].last_seen = datetime.now(timezone.utc)

    def search_network(self, query: str) -> list[tuple[str, SharedFile]]:
        """Search all known peers' catalogs for files matching the query."""
        results = []
        q = query.lower()
        for peer_id, peer in self.peers.items():
            for sf in peer.shared_files:
                if q in sf.name.lower():
                    results.append((peer_id, sf))
        return results

    # --- File transfer ---

    def request_file(self, provider_id: str, filename: str) -> Optional[TransferRequest]:
        """Initiate a file transfer request from a peer."""
        peer = self.peers.get(provider_id)
        if peer is None:
            return None
        sf = next((f for f in peer.shared_files if f.name == filename), None)
        if sf is None:
            return None

        self._next_request_id += 1
        req_id = f"xfer-{self._next_request_id}"
        req = TransferRequest(
            request_id=req_id,
            filename=filename,
            requester_id=self.peer_id,
            provider_id=provider_id,
            total_chunks=sf.chunk_count,
            started=datetime.now(timezone.utc),
        )
        req.status = "active"
        self.transfers[req_id] = req
        self._chunks_store[req_id] = [None] * sf.chunk_count
        return req

    def receive_chunk(self, request_id: str, chunk: FileChunk) -> bool:
        """Receive a chunk for an active transfer. Returns True if chunk is valid."""
        req = self.transfers.get(request_id)
        if req is None or req.status != "active":
            return False
        if not chunk.verify():
            return False
        chunks = self._chunks_store.get(request_id)
        if chunks is None or chunk.index >= len(chunks):
            return False
        chunks[chunk.index] = chunk
        req.chunks_received = sum(1 for c in chunks if c is not None)
        if req.chunks_received == req.total_chunks:
            req.status = "completed"
            req.completed = datetime.now(timezone.utc)
            # Reassemble file
            assembled = b"".join(c.data for c in chunks if c is not None)
            self.local_files[req.filename] = assembled
        return True

    def get_missing_chunks(self, request_id: str) -> list[int]:
        """Return indices of chunks not yet received for a transfer."""
        chunks = self._chunks_store.get(request_id)
        if chunks is None:
            return []
        return [i for i, c in enumerate(chunks) if c is None]

    # --- Network-wide stats ---

    def network_stats(self) -> dict:
        """Summary of this peer's view of the network."""
        total_files = sum(len(p.shared_files) for p in self.peers.values())
        total_bytes = sum(
            sf.size for p in self.peers.values() for sf in p.shared_files
        )
        return {
            "peer_id": self.peer_id,
            "known_peers": len(self.peers),
            "local_files": len(self.local_files),
            "network_files": total_files,
            "network_bytes": total_bytes,
            "active_transfers": sum(
                1 for t in self.transfers.values() if t.status == "active"
            ),
        }


def simulate_transfer(provider: PeerNode, requester: PeerNode, filename: str) -> Optional[TransferRequest]:
    """Simulate a complete file transfer between two peers in-process."""
    req = requester.request_file(provider.peer_id, filename)
    if req is None:
        return None
    for idx in range(req.total_chunks):
        chunk = provider.get_chunk(filename, idx)
        if chunk is not None:
            requester.receive_chunk(req.request_id, chunk)
    return req


# --- demo ---
if __name__ == "__main__":
    # Create two peers
    alice = PeerNode("alice", "192.168.1.10", 8001)
    bob = PeerNode("bob", "192.168.1.11", 8002)

    # Mutual discovery
    alice.discover_peer("bob", "192.168.1.11", 8002)
    bob.discover_peer("alice", "192.168.1.10", 8001)

    # Alice shares a file
    content = b"The Rosetta Stone is not a learning exercise." * 100
    alice.add_file("rosetta.txt", content)

    # Exchange catalogs
    bob.receive_catalog("alice", alice.get_catalog())

    # Bob searches and finds it
    results = bob.search_network("rosetta")
    print(f"Search results: {[(pid, sf.name, sf.size) for pid, sf in results]}")

    # Bob requests the file
    req = simulate_transfer(alice, bob, "rosetta.txt")
    if req:
        print(f"Transfer {req.request_id}: {req.status} ({req.progress_pct})")
        print(f"Bob now has {len(bob.local_files)} local file(s)")
        print(f"File integrity: {bob.local_files['rosetta.txt'] == content}")

File: go/networking/p2p_file_sharing.go
// P2P File Sharing App — decentralized file exchange between peers.
package networking

import (
	"crypto/sha256"
	"fmt"
	"strings"
	"sync"
	"time"
)

// FileChunk is a chunk of a file with integrity verification.
type FileChunk struct {
	Index  int
	Data   []byte
	Sha256 string
}

// NewFileChunk creates a chunk with a computed hash.
func NewFileChunk(index int, data []byte) FileChunk {
	return FileChunk{
		Index:  index,
		Data:   append([]byte(nil), data...),
		Sha256: fmt.Sprintf("%x", sha256.Sum256(data)),
	}
}

// Verify checks that the chunk data matches its hash.
func (c FileChunk) Verify() bool {
	return fmt.Sprintf("%x", sha256.Sum256(c.Data)) == c.Sha256
}

// SharedFile describes a file available from a peer.
type SharedFile struct {
	Name       string
	Size       int
	Sha256     string
	ChunkSize  int
	ChunkCount int
}

// NewSharedFile creates file metadata with computed chunk count.
func NewSharedFile(name string, size int, sha string, chunkSize int) SharedFile {
	count := 0
	if size > 0 {
		count = (size + chunkSize - 1) / chunkSize
	}
	return SharedFile{Name: name, Size: size, Sha256: sha, ChunkSize: chunkSize, ChunkCount: count}
}

// P2PPeerInfo describes a known peer.
type P2PPeerInfo struct {
	PeerID      string
	Address     string
	Port        int
	LastSeen    time.Time
	SharedFiles []SharedFile
}

// IsStale returns true if the peer hasn't been seen within timeout.
func (p *P2PPeerInfo) IsStale(timeout time.Duration) bool {
	return time.Since(p.LastSeen) > timeout
}

// TransferStatus tracks the state of a file transfer.
type P2PTransferStatus int

const (
	P2PTransferPending P2PTransferStatus = iota
	P2PTransferActive
	P2PTransferCompleted
	P2PTransferFailed
)

// P2PTransferRequest tracks a file transfer between peers.
type P2PTransferRequest struct {
	RequestID      string
	Filename       string
	RequesterID    string
	ProviderID     string
	Status         P2PTransferStatus
	ChunksReceived int
	TotalChunks    int
}

// Progress returns transfer completion as a float [0, 1].
func (t *P2PTransferRequest) Progress() float64 {
	if t.TotalChunks == 0 {
		return 0
	}
	return float64(t.ChunksReceived) / float64(t.TotalChunks)
}

// PeerNode is a peer in the P2P network.
type PeerNode struct {
	mu            sync.RWMutex
	PeerID        string
	Address       string
	Port          int
	peers         map[string]*P2PPeerInfo
	localFiles    map[string][]byte
	sharedCatalog map[string]SharedFile
	transfers     map[string]*P2PTransferRequest
	chunksStore   map[string][]*FileChunk
	nextReqID     int
}

// NewPeerNode creates a new peer.
func NewPeerNode(peerID, address string, port int) *PeerNode {
	return &PeerNode{
		PeerID:        peerID,
		Address:       address,
		Port:          port,
		peers:         make(map[string]*P2PPeerInfo),
		localFiles:    make(map[string][]byte),
		sharedCatalog: make(map[string]SharedFile),
		transfers:     make(map[string]*P2PTransferRequest),
		chunksStore:   make(map[string][]*FileChunk),
	}
}

// AddFile adds a file to the local store and shares it.
func (n *PeerNode) AddFile(name string, data []byte, chunkSize int) SharedFile {
	n.mu.Lock()
	defer n.mu.Unlock()
	hash := fmt.Sprintf("%x", sha256.Sum256(data))
	cp := make([]byte, len(data))
	copy(cp, data)
	n.localFiles[name] = cp
	sf := NewSharedFile(name, len(data), hash, chunkSize)
	n.sharedCatalog[name] = sf
	return sf
}

// RemoveFile stops sharing a file.
func (n *PeerNode) RemoveFile(name string) {
	n.mu.Lock()
	defer n.mu.Unlock()
	delete(n.localFiles, name)
	delete(n.sharedCatalog, name)
}

// GetChunk returns a specific chunk of a local file.
func (n *PeerNode) GetChunk(filename string, index int) (*FileChunk, bool) {
	n.mu.RLock()
	defer n.mu.RUnlock()
	data, ok := n.localFiles[filename]
	if !ok {
		return nil, false
	}
	sf, ok := n.sharedCatalog[filename]
	if !ok {
		return nil, false
	}
	start := index * sf.ChunkSize
	if start >= len(data) {
		return nil, false
	}
	end := start + sf.ChunkSize
	if end > len(data) {
		end = len(data)
	}
	chunk := NewFileChunk(index, data[start:end])
	return &chunk, true
}

// DiscoverPeer registers a known peer.
func (n *PeerNode) DiscoverPeer(peerID, address string, port int) {
	n.mu.Lock()
	defer n.mu.Unlock()
	n.peers[peerID] = &P2PPeerInfo{
		PeerID:   peerID,
		Address:  address,
		Port:     port,
		LastSeen: time.Now().UTC(),
	}
}

// RemovePeer removes a peer.
func (n *PeerNode) RemovePeer(peerID string) {
	n.mu.Lock()
	defer n.mu.Unlock()
	delete(n.peers, peerID)
}

// PruneStalePeers removes peers not seen within timeout.
func (n *PeerNode) PruneStalePeers(timeout time.Duration) []string {
	n.mu.Lock()
	defer n.mu.Unlock()
	var stale []string
	for id, p := range n.peers {
		if p.IsStale(timeout) {
			stale = append(stale, id)
		}
	}
	for _, id := range stale {
		delete(n.peers, id)
	}
	return stale
}

// Heartbeat updates last-seen for a peer.
func (n *PeerNode) Heartbeat(peerID string) {
	n.mu.Lock()
	defer n.mu.Unlock()
	if p, ok := n.peers[peerID]; ok {
		p.LastSeen = time.Now().UTC()
	}
}

// GetCatalog returns this peer's shared files.
func (n *PeerNode) GetCatalog() []SharedFile {
	n.mu.RLock()
	defer n.mu.RUnlock()
	catalog := make([]SharedFile, 0, len(n.sharedCatalog))
	for _, sf := range n.sharedCatalog {
		catalog = append(catalog, sf)
	}
	return catalog
}

// ReceiveCatalog stores another peer's catalog.
func (n *PeerNode) ReceiveCatalog(peerID string, catalog []SharedFile) {
	n.mu.Lock()
	defer n.mu.Unlock()
	if p, ok := n.peers[peerID]; ok {
		p.SharedFiles = catalog
		p.LastSeen = time.Now().UTC()
	}
}

// SearchNetwork finds files matching query across all known peers.
func (n *PeerNode) SearchNetwork(query string) []struct {
	PeerID string
	File   SharedFile
} {
	n.mu.RLock()
	defer n.mu.RUnlock()
	q := strings.ToLower(query)
	var results []struct {
		PeerID string
		File   SharedFile
	}
	for pid, peer := range n.peers {
		for _, sf := range peer.SharedFiles {
			if strings.Contains(strings.ToLower(sf.Name), q) {
				results = append(results, struct {
					PeerID string
					File   SharedFile
				}{pid, sf})
			}
		}
	}
	return results
}

// RequestFile initiates a transfer request. Returns the request ID.
func (n *PeerNode) RequestFile(providerID, filename string) (string, error) {
	n.mu.Lock()
	defer n.mu.Unlock()
	peer, ok := n.peers[providerID]
	if !ok {
		return "", fmt.Errorf("unknown peer: %s", providerID)
	}
	var sf *SharedFile
	for i := range peer.SharedFiles {
		if peer.SharedFiles[i].Name == filename {
			sf = &peer.SharedFiles[i]
			break
		}
	}
	if sf == nil {
		return "", fmt.Errorf("file %q not in peer %s catalog", filename, providerID)
	}
	n.nextReqID++
	reqID := fmt.Sprintf("xfer-%d", n.nextReqID)
	n.transfers[reqID] = &P2PTransferRequest{
		RequestID:   reqID,
		Filename:    filename,
		RequesterID: n.PeerID,
		ProviderID:  providerID,
		Status:      P2PTransferActive,
		TotalChunks: sf.ChunkCount,
	}
	chunks := make([]*FileChunk, sf.ChunkCount)
	n.chunksStore[reqID] = chunks
	return reqID, nil
}

// ReceiveChunk accepts a chunk for an active transfer.
func (n *PeerNode) ReceiveChunk(requestID string, chunk FileChunk) bool {
	n.mu.Lock()
	defer n.mu.Unlock()
	req, ok := n.transfers[requestID]
	if !ok || req.Status != P2PTransferActive {
		return false
	}
	if !chunk.Verify() {
		return false
	}
	chunks := n.chunksStore[requestID]
	if chunk.Index >= len(chunks) {
		return false
	}
	chunks[chunk.Index] = &chunk
	received := 0
	for _, c := range chunks {
		if c != nil {
			received++
		}
	}
	req.ChunksReceived = received
	if received == req.TotalChunks {
		req.Status = P2PTransferCompleted
		var assembled []byte
		for _, c := range chunks {
			if c != nil {
				assembled = append(assembled, c.Data...)
			}
		}
		n.localFiles[req.Filename] = assembled
	}
	return true
}

// GetTransfer returns the status of a transfer.
func (n *PeerNode) GetTransfer(requestID string) (*P2PTransferRequest, bool) {
	n.mu.RLock()
	defer n.mu.RUnlock()
	r, ok := n.transfers[requestID]
	return r, ok
}

// HasFile checks if a file exists locally.
func (n *PeerNode) HasFile(name string) bool {
	n.mu.RLock()
	defer n.mu.RUnlock()
	_, ok := n.localFiles[name]
	return ok
}

// SimulateP2PTransfer runs a complete transfer between two peers in-process.
func SimulateP2PTransfer(provider, requester *PeerNode, filename string) (string, error) {
	reqID, err := requester.RequestFile(provider.PeerID, filename)
	if err != nil {
		return "", err
	}
	req, _ := requester.GetTransfer(reqID)
	for idx := 0; idx < req.TotalChunks; idx++ {
		chunk, ok := provider.GetChunk(filename, idx)
		if ok {
			requester.ReceiveChunk(reqID, *chunk)
		}
	}
	return reqID, nil
}
